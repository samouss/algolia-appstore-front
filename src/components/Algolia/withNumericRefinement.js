import React, { Component } from 'react';
import { getDisplayName } from 'core/utils';
import { ContextTypes } from './Provider';

const withNumericRefinement = ({
  attribute,
  operator,
} = {}) => WrappedComponent => {
  if (!attribute) {
    throw new Error('You must provide the attribute parameter.');
  }

  if (!operator) {
    throw new Error('You must provide the operator parameter.');
  }

  class WithDisjunctiveFacetRefinement extends Component {

    constructor(props) {
      super(props);

      this.onChange = this.onChange.bind(this);
    }

    onChange(value, isRefined) {
      const { algoliaHelper } = this.context;

      if (isRefined) {
        return algoliaHelper
          .removeNumericRefinement(attribute)
          .search();
      }

      algoliaHelper
        .removeNumericRefinement(attribute)
        .addNumericRefinement(attribute, operator, value)
        .search();
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          onChange={this.onChange}
        />
      );
    }

  }

  WithDisjunctiveFacetRefinement.displayName = getDisplayName(
    WrappedComponent,
    'withNumericRefinement',
  );

  WithDisjunctiveFacetRefinement.contextTypes = ContextTypes;

  return WithDisjunctiveFacetRefinement;
};

export default withNumericRefinement;
