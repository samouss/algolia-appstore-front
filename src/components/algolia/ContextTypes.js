import PropTypes from 'prop-types';

export default {
  algoliaClient: PropTypes.object.isRequired,
  algoliaHelper: PropTypes.object.isRequired,
};
