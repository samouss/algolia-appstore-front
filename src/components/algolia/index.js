export { default as Provider } from './Provider';
export { default as withPaginateResults } from './withPaginateResults';
