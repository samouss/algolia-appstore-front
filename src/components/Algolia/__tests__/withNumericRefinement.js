import React from 'react';
import { shallow } from 'enzyme';
import { createMockAlgoliaClient, createMockAlgoliaHelper } from 'test/algolia';
import withNumericRefinement from '../withNumericRefinement';

describe('algolia', () => {
  describe('withNumericRefinement', () => {
    const createContext = () => ({
      algoliaHelper: createMockAlgoliaHelper(),
      algoliaClient: createMockAlgoliaClient(),
    });

    const defaultParameters = {
      attribute: 'rating',
      operator: '>=',
    };

    const Component = props => (
      <input {...props} />
    );

    it('expect to render', () => {
      const context = createContext();

      const parameters = {
        ...defaultParameters,
      };

      const props = {
        className: 'sample-class-name',
      };

      const ApplyComponent = withNumericRefinement(parameters)(Component);

      const component = shallow(
        <ApplyComponent
          {...props}
        />,
        { context },
      );

      expect(component).toMatchSnapshot();
    });

    it('expect to throw when attribute is not provided', () => {
      expect(() => withNumericRefinement()(Component)).toThrow(
        'You must provide the attribute parameter.',
      );
    });

    it('expect to throw when operator is not provided', () => {
      expect(() => withNumericRefinement({ attribute: 'rating' })(Component)).toThrow(
        'You must provide the operator parameter.',
      );
    });

    it('expect to call onChange', () => {
      const context = createContext();

      const parameters = {
        ...defaultParameters,
      };

      const ApplyComponent = withNumericRefinement(parameters)(Component);

      const onChange = jest.spyOn(ApplyComponent.prototype, 'onChange');

      const component = shallow(
        <ApplyComponent />,
        { context },
      );

      component
        .find('Component')
        .simulate('change');

      expect(onChange).toHaveBeenCalled();
      onChange.mockRestore();
    });

    it('expect to have correct displayName', () => {
      const parameters = {
        ...defaultParameters,
      };

      const ApplyComponent = withNumericRefinement(parameters)(Component);

      const expectation = 'withNumericRefinement(Component)';
      const actual = ApplyComponent.displayName;

      expect(actual).toBe(expectation);
    });

    describe('onChange', () => {
      it('expect to call removeNumericRefinement and search when isRefined', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withNumericRefinement(parameters)(Component);

        const component = shallow(
          <ApplyComponent />,
          { context },
        );

        component.instance().onChange(
          3,
          true,
        );

        expect(context.algoliaHelper.removeNumericRefinement).toHaveBeenCalledWith(
          parameters.attribute,
        );

        expect(context.algoliaHelper.addNumericRefinement).not.toHaveBeenCalled();

        expect(context.algoliaHelper.search).toHaveBeenCalled();
      });

      it('expect to call removeDisjunctive, addNumericRefinement and search', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withNumericRefinement(parameters)(Component);

        const component = shallow(
          <ApplyComponent />,
          { context },
        );

        component.instance().onChange(
          3,
          false,
        );

        expect(context.algoliaHelper.removeNumericRefinement).toHaveBeenCalledWith(
          parameters.attribute,
        );

        expect(context.algoliaHelper.addNumericRefinement).toHaveBeenCalledWith(
          parameters.attribute,
          parameters.operator,
          3,
        );

        expect(context.algoliaHelper.search).toHaveBeenCalled();
      });
    });
  });
});
