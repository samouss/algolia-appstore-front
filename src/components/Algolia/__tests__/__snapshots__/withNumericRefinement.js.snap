// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`algolia withNumericRefinement expect to render 1`] = `
<Component
  className="sample-class-name"
  onChange={[Function]}
/>
`;
